source/reduce/structured_loop_to_selection_reduction_opportunity.cpp

```diff
@@ -0,0 +1,406 @@
+      ->IsReachable(loop_construct_header_);
+}
+
+void StructuredLoopToSelectionReductionOpportunity::Apply() {
+  // Force computation of dominator analysis and CFG before we start to mess
+  // with edges in the function.
+  context_->GetDominatorAnalysis(enclosing_function_);
+  context_->cfg();
+
+  // (1) Redirect edges that point to the loop's continue target to their
+  // closest merge block.
+  RedirectToClosestMergeBlock(
+      loop_construct_header_->GetLoopMergeInst()->GetSingleWordOperand(
+          kContinueNodeIndex));
+
+  // (2) Redirect edges that point to the loop's merge block to their closest
+  // merge block (which might be that of an enclosing selection, for instance).
+  RedirectToClosestMergeBlock(
+      loop_construct_header_->GetLoopMergeInst()->GetSingleWordOperand(
+          kMergeNodeIndex));
+
+  // (3) Turn the loop construct header into a selection.
+  ChangeLoopToSelection();
+
+  // We have made control flow changes that do not preserve the analyses that
+  // were performed.
+  context_->InvalidateAnalysesExceptFor(IRContext::Analysis::kAnalysisNone);
+
+  // (4) By changing CFG edges we may have created scenarios where ids are used
+  // without being dominated; we fix instances of this.
+  FixNonDominatedIdUses();
+
+  // Invalidate the analyses we just used.
+  context_->InvalidateAnalysesExceptFor(IRContext::Analysis::kAnalysisNone);
+}
+
+void StructuredLoopToSelectionReductionOpportunity::RedirectToClosestMergeBlock(
+    uint32_t original_target_id) {
+  // Consider every predecessor of the node with respect to which edges should
+  // be redirected.
+  for (auto pred : context_->cfg()->preds(original_target_id)) {
+    if (!context_->GetDominatorAnalysis(enclosing_function_)
+             ->IsReachable(pred)) {
+      // We do not care about unreachable predecessors (and dominance
+      // information, and thus the notion of structured control flow, makes
+      // little sense for unreachable blocks).
+      continue;
+    }
+    // Find the merge block of the structured control construct that most
+    // tightly encloses the predecessor.
+    auto new_merge_target = FindClosestMerge(pred);
+    assert(new_merge_target != pred);
+
+    if (!new_merge_target) {
+      // If the loop being transformed is outermost, and the predecessor is
+      // part of that loop's continue construct, there will be no such
+      // enclosing control construct.  In this case, the continue construct
+      // will become unreachable anyway, so it is fine not to redirect the
+      // edge.
+      continue;
+    }
+
+    if (new_merge_target != original_target_id) {
+      // Redirect the edge if it doesn't already point to the desired block.
+      RedirectEdge(pred, original_target_id, new_merge_target);
+    }
+  }
+}
+
+uint32_t StructuredLoopToSelectionReductionOpportunity::FindClosestMerge(
+    uint32_t block_id) {
+  // We want to find the merge block associated with the structured control flow
```

Comment:
It would be great to use StructuredCFGAnalysis here if possible.  The loop header issue (which also seems to apply to selection headers) is easy to work around.

However, there appear to be two further ways in which this analysis doesn't follow the SPIR-V spec to the letter of the law:

1. Unreachable code might be regarded as part of a construct, but unreachable code isn't dominated by anything, including the construct's header, so according to the spec can't be part of the construct.

2. Continue constructs are regarded as being part of the loop whose header they are associated with.  This likely makes sense in many contexts, but according to the spec the loop construct is different from the continue construct and doesn't contain it.  I believe it's even OK for the continue construct to occur after the loop merge block in program order, and I worry that this could cause problems sometimes.

Currently what I have implemented does try to follow the SPIR-V spec accurately in these regards.  I need to think more on whether the correctness of the transformation can rely on this, or whether if doesn't really matter.

Reply:
I have reworked things to use structured CFG analysis.  This required changing a few tests due to differences in how that analysis works compared with what I had implemented; the test changes are to do with unreachable code and continue targets, and all still lead to valid SPIR-V.